source/extensions/filters/http/jwt_authn/filter.cc

```diff
@@ -56,12 +65,35 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
     return Http::FilterHeadersStatus::Continue;
   }
 
-  // Verify the JWT token, onComplete() will be called when completed.
-  const auto* verifier =
-      config_->findVerifier(headers, *decoder_callbacks_->streamInfo().filterState());
-  if (!verifier) {
+  const Verifier* verifier = nullptr;
+  bool use_per_route = false;
+  Router::RouteConstSharedPtr route = decoder_callbacks_->route();
+  if (route != nullptr && route->routeEntry() != nullptr) {
+    const auto* per_route_config =
+        Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteFilterConfig>(
+            HttpFilterNames::get().JwtAuthn, route);
+    if (per_route_config != nullptr) {
+      use_per_route = true;
+      std::string error_msg;
+      std::tie(verifier, error_msg) = config_->findPerRouteVerifier(*per_route_config);
+      if (!error_msg.empty()) {
+        stats_.denied_.inc();
+        state_ = Responded;
+        decoder_callbacks_->sendLocalReply(Http::Code::Forbidden,
+                                           absl::StrCat("Failed JWT authentication: ", error_msg),
+                                           nullptr, absl::nullopt, generateRcDetails(error_msg));
+        return Http::FilterHeadersStatus::StopIteration;
+      }
+    }
+  }
+  if (!use_per_route) {
+    verifier = config_->findVerifier(headers, *decoder_callbacks_->streamInfo().filterState());
+  }
```

Comment:
extract this part to a method?

Reply:
No need, after remove if condition, it is very simple now.